### PR TITLE
Fix error being returned from query exp even when enclosed within on fail

### LINKED ISSUE #34894: Compiler crashes when check expressions are used in from queries when the error would be caught by an on fail clause
**Description:**
A from query containing a check expression can trip up the compiler when the error is not directly returned by the surrounding function but caught by an `on fail` clause of a `do` or a `transaction` block.

**Steps to reproduce:**

Try to compile the minimal example below:

```ballerina
import ballerina/io;
public function main() {
    do {
        _ = from string v in []
            select check someFunction(v);
    } on fail var err {
        io:println(err);
    }
}

function someFunction(string input) returns string|error {
    return input;
}
```

<details>

<summary>
Click here to show more information.
</summary>

Stacktrace:
```text
java.lang.NullPointerException
        at org.wso2.ballerinalang.compiler.bir.codegen.JvmTerminatorGen.genGoToTerm(JvmTerminatorGen.java:386)
        at org.wso2.ballerinalang.compiler.bir.codegen.JvmTerminatorGen.genTerminator(JvmTerminatorGen.java:329)
        at org.wso2.ballerinalang.compiler.bir.codegen.methodgen.MethodGen.generateBasicBlocks(MethodGen.java:486)
        at org.wso2.ballerinalang.compiler.bir.codegen.methodgen.MethodGen.genJMethodForBFunc(MethodGen.java:235)
        at org.wso2.ballerinalang.compiler.bir.codegen.methodgen.MethodGen.generateMethod(MethodGen.java:161)
        at org.wso2.ballerinalang.compiler.bir.codegen.JvmPackageGen.lambda$generateModuleClasses$0(JvmPackageGen.java:446)
        at java.base/java.util.HashMap$EntrySet.forEach(HashMap.java:1039)
        at org.wso2.ballerinalang.compiler.bir.codegen.JvmPackageGen.generateModuleClasses(JvmPackageGen.java:397)
        at org.wso2.ballerinalang.compiler.bir.codegen.JvmPackageGen.generate(JvmPackageGen.java:811)
        at org.wso2.ballerinalang.compiler.bir.codegen.CodeGenerator.generate(CodeGenerator.java:95)
        at org.wso2.ballerinalang.compiler.bir.codegen.CodeGenerator.generate(CodeGenerator.java:70)
        at io.ballerina.projects.JBallerinaBackend.performCodeGen(JBallerinaBackend.java:301)
        at io.ballerina.projects.ModuleContext.generateCodeInternal(ModuleContext.java:466)
        at io.ballerina.projects.ModuleCompilationState$4.generatePlatformSpecificCode(ModuleCompilationState.java:132)
        at io.ballerina.projects.ModuleContext.generatePlatformSpecificCode(ModuleContext.java:389)
        at io.ballerina.projects.JBallerinaBackend.performCodeGen(JBallerinaBackend.java:159)
        at io.ballerina.projects.JBallerinaBackend.<init>(JBallerinaBackend.java:137)
        at io.ballerina.projects.JBallerinaBackend.lambda$from$0(JBallerinaBackend.java:113)
        at java.base/java.util.HashMap.computeIfAbsent(HashMap.java:1133)
        at io.ballerina.projects.PackageCompilation.getCompilerBackend(PackageCompilation.java:182)
        at io.ballerina.projects.JBallerinaBackend.from(JBallerinaBackend.java:112)
        at io.ballerina.cli.task.CompileTask.execute(CompileTask.java:93)
        at io.ballerina.cli.TaskExecutor.executeTasks(TaskExecutor.java:40)
        at io.ballerina.cli.cmd.BuildCommand.execute(BuildCommand.java:249)
        at java.base/java.util.Optional.ifPresent(Optional.java:183)
        at io.ballerina.cli.launcher.Main.main(Main.java:51)
```

The following code fails with a different stacktrace:

```ballerina
import ballerina/io;
public function main() {
    transaction { // this leads to a different stacktrace
        _ = check from string v in []
            do { _ = check someFunction(v); }; // just another way to trigger the bug?
        check commit;
    } on fail var err {
        io:println(err);
    }
}

function someFunction(string input) returns string|error {
    return input;
}
```

Stacktrace:
```text
java.lang.NullPointerException
        at org.wso2.ballerinalang.compiler.bir.optimizer.BIROptimizer$RHSTempVarOptimizer.visit(BIROptimizer.java:164)
        at org.wso2.ballerinalang.compiler.bir.model.BIRNode$BIRBasicBlock.accept(BIRNode.java:443)
        at org.wso2.ballerinalang.compiler.bir.optimizer.BIROptimizer$RHSTempVarOptimizer.lambda$visit$4(BIROptimizer.java:126)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
        at org.wso2.ballerinalang.compiler.bir.optimizer.BIROptimizer$RHSTempVarOptimizer.visit(BIROptimizer.java:126)
        at org.wso2.ballerinalang.compiler.bir.model.BIRNode$BIRFunction.accept(BIRNode.java:401)
        at org.wso2.ballerinalang.compiler.bir.optimizer.BIROptimizer$RHSTempVarOptimizer.lambda$visit$1(BIROptimizer.java:103)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
        at org.wso2.ballerinalang.compiler.bir.optimizer.BIROptimizer$RHSTempVarOptimizer.visit(BIROptimizer.java:103)
        at org.wso2.ballerinalang.compiler.bir.model.BIRNode$BIRPackage.accept(BIRNode.java:83)
        at org.wso2.ballerinalang.compiler.bir.optimizer.BIROptimizer.optimizePackage(BIROptimizer.java:82)
        at org.wso2.ballerinalang.compiler.bir.BIRGen.genBIR(BIRGen.java:269)
        at io.ballerina.projects.internal.CompilerPhaseRunner.birGen(CompilerPhaseRunner.java:220)
        at io.ballerina.projects.internal.CompilerPhaseRunner.performBirGenPhases(CompilerPhaseRunner.java:157)
        at io.ballerina.projects.ModuleContext.generateCodeInternal(ModuleContext.java:452)
        at io.ballerina.projects.ModuleCompilationState$4.generatePlatformSpecificCode(ModuleCompilationState.java:132)
        at io.ballerina.projects.ModuleContext.generatePlatformSpecificCode(ModuleContext.java:389)
        at io.ballerina.projects.JBallerinaBackend.performCodeGen(JBallerinaBackend.java:159)
        at io.ballerina.projects.JBallerinaBackend.<init>(JBallerinaBackend.java:137)
        at io.ballerina.projects.JBallerinaBackend.lambda$from$0(JBallerinaBackend.java:113)
        at java.base/java.util.HashMap.computeIfAbsent(HashMap.java:1133)
        at io.ballerina.projects.PackageCompilation.getCompilerBackend(PackageCompilation.java:182)
        at io.ballerina.projects.JBallerinaBackend.from(JBallerinaBackend.java:112)
        at io.ballerina.cli.task.CompileTask.execute(CompileTask.java:93)
        at io.ballerina.cli.TaskExecutor.executeTasks(TaskExecutor.java:40)
        at io.ballerina.cli.cmd.BuildCommand.execute(BuildCommand.java:249)
        at java.base/java.util.Optional.ifPresent(Optional.java:183)
        at io.ballerina.cli.launcher.Main.main(Main.java:51)
```

The `transaction` block causes the error even when the `on fail` clause is missing:

```ballerina
public function main() returns error? {
    transaction {
        _ = check from string v in []
            do {_ = check someFunction(v);};
    } 
}
```

But changing it to a `do` block will compile just fine.

</details>

Possible Workaround until the issue is resolved:

```ballerina
// trap panic, convert back to error
_ = check trap from string v in []
    select checkpanic someFunction(v); // panic on error
```

**Affected Versions:**

Tested on `slbeta6` but the error probably exists since `slbeta4`. Similar code has compiled fine with `slbeta3`.

**OS, DB, other environment details and versions:**

Ubuntu (on WSL2)

**Related Issues:**

Probably related: #34557 

There are more issues with similar stacktraces of compiler errors.

**Suggested Labels:**

Team/CompilerFE, Area/BIR?

### PR DESCRIPTION
**Pull request by pcnfernando**
Fixes #34894

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

**Comment by gimantha** [comment]
We need to change the base @pcnfernando

File: tests/jballerina-unit-test/src/test/resources/test-src/query/query_expr_with_errors.bal [hunk 1]
```diff
@@ -210,6 +210,50 @@ function throwUnreachableError() returns int|UnreachableError {
...
+
+function testErrorReturnedFromSelect() {
+    assertTrue(checkErrorAtSelect() is error);
+}
+
+function checkErrorAtSelect() returns error? {
+    _ = from int v in 1 ... 3
+        select check verifyCheck(v);
+}
+
```
Comment (lasinicl) [inline tests/jballerina-unit-test/src/test/resources/test-src/query/query_expr_with_errors.bal:256]: Doesn't work with `where` clause?
The following crashes.
```ballerina
import ballerina/io;
public function main() returns error? {
    do {
        var ans = from int v in 1 ... 3
            where v == check verifyCheck(1)
            select v;

        io:println(ans);
    } on fail var err {
        io:println(err);
    }
}

public function verifyCheck(int i) returns int|error {
    return error("Verify Check.");
}
```

Shall we also add some tests with nested queries?
Eg:
```ballerina
public function main() returns error? {
    do {
        _ = from int a in (from int v in [1, 2] let int m = check verifyCheck(v) select m)
            select a;
    } on fail var err {
        io:println(err);
    }
}
```

Comment (pcnfernando) [reply tests/jballerina-unit-test/src/test/resources/test-src/query/query_expr_with_errors.bal:256]: Seems to be an issue with how we handle the `where` vs `let` clause during QueryDesugar. The let version works. Will check
```ballerina
function dummyTest() {
    do {
            var ans = from int v in 1 ... 3
                let boolean bool = v == check verifyCheck(1)
                where bool
                select v;

            print(ans);
        } on fail var err {
            print(err);
        }
}

public function verifyCheck(int i) returns int|error {
    return error("Verify Check.");
}

```

Comment (pcnfernando) [reply tests/jballerina-unit-test/src/test/resources/test-src/query/query_expr_with_errors.bal:256]: Created an issue https://github.com/ballerina-platform/ballerina-lang/issues/35885 to handle this separately as the complete pipeline changes when the return type of the lambda function which is used for wrapping where clause changed from `boolean` to `error|boolean`

**Review by lasinicl — APPROVED** [review]

**Pull request merged**